perception/map_based_prediction/src/map_based_prediction_node.cpp

```diff
@@ -1119,25 +1231,84 @@ std::vector<PredictedRefPath> MapBasedPredictionNode::getPredictedReferencePath(
     const double search_dist = prediction_time_horizon_ * obj_vel +
                                lanelet::utils::getLaneletLength3d(current_lanelet_data.lanelet);
     lanelet::routing::PossiblePathsParams possible_params{search_dist, {}, 0, false, true};
+    const double validate_time_horizon =
+      prediction_time_horizon_ * prediction_time_horizon_rate_for_validate_lane_length_;
 
     // Step1. Get the path
     // Step1.1 Get the left lanelet
     lanelet::routing::LaneletPaths left_paths;
     auto opt_left = routing_graph_ptr_->left(current_lanelet_data.lanelet);
+    auto adjacent_left = routing_graph_ptr_->adjacentLeft(current_lanelet_data.lanelet);
     if (!!opt_left) {
       left_paths = routing_graph_ptr_->possiblePaths(*opt_left, possible_params);
+    } else if (!!adjacent_left) {
+      left_paths = routing_graph_ptr_->possiblePaths(*adjacent_left, possible_params);
+    } else {
+      // if no left lanelet in graph search for unconnected lanelet
+      auto unconnected_lefts =
+        getLeftLineSharingLanelets(current_lanelet_data.lanelet, lanelet_map_ptr_);
+      // search for only first unconnected lanelet
+      for (const auto & unconnected_left_lanelet : unconnected_lefts) {
+        // if lanelet is isolated in routing graph
+        if (isIsolatedLanelet(unconnected_left_lanelet, routing_graph_ptr_)) {
+          // need to validate isolated lanelet has enough length
+          if (!validateIsolatedLaneletLength(
+                unconnected_left_lanelet, object, validate_time_horizon))
+            continue;
+          // get possible path for isolated lanelet
+          left_paths = getPossiblePathsForIsolatedLanelet(unconnected_left_lanelet);
+        } else {
+          // if it has connection it should have paths
+          left_paths = routing_graph_ptr_->possiblePaths(unconnected_left_lanelet, possible_params);
+          if (!left_paths.empty()) break;
+        }
+      }
     }
```

Comment:
just a suggestion you can ignore

consolidate implementation because it seems getting right and left lanelets are the same, 
if you use the lammda, the code can be like this?
```    
    // Step1.1 and Step1.2 Get the lanelets
    auto getLaneletPaths = [&](auto get_direction, auto get_line_sharing, auto maneuver) {
      lanelet::routing::LaneletPaths paths;
      auto opt_dir = get_direction(current_lanelet_data.lanelet);
      auto adjacent_dir = get_direction == routing_graph_ptr_->right
                            ? routing_graph_ptr_->adjacentRight(current_lanelet_data.lanelet)
                            : routing_graph_ptr_->adjacentLeft(current_lanelet_data.lanelet);
      if (!!opt_dir) {
        paths = routing_graph_ptr_->possiblePaths(*opt_dir, possible_params);
      } else if (!!adjacent_dir) {
        paths = routing_graph_ptr_->possiblePaths(*adjacent_dir, possible_params);
      } else {
        auto unconnected_dir = get_line_sharing(current_lanelet_data.lanelet, lanelet_map_ptr_);
        for (const auto & unconnected_lanelet : unconnected_dir) {
          if (isIsolatedLanelet(unconnected_lanelet, routing_graph_ptr_)) {
            if (!validateIsolatedLaneletLength(unconnected_lanelet, object, validate_time_horizon))
              continue;
            paths = getPossiblePathsForIsolatedLanelet(unconnected_lanelet);
          } else {
            paths = routing_graph_ptr_->possiblePaths(unconnected_lanelet, possible_params);
            if (!paths.empty()) break;
          }
        }
      }
      return paths;
    };

    lanelet::routing::LaneletPaths left_paths = getLaneletPaths(
      routing_graph_ptr_->left, getLeftLineSharingLanelets, Maneuver::LEFT_LANE_CHANGE);
    lanelet::routing::LaneletPaths right_paths = getLaneletPaths(
      routing_graph_ptr_->right, getRightLineSharingLanelets, Maneuver::RIGHT_LANE_CHANGE);
```

Reply:
Your suggestion make sense.
I think it is better to separate neighbor lanelets detection process and paths creation process, so added following changes:

- https://github.com/autowarefoundation/autoware.universe/pull/4353/commits/e3762cc4e2208963085a9e769892f0041be04240 : paths estimation to lambda function
- https://github.com/autowarefoundation/autoware.universe/pull/4353/commits/b69b3b38e85673199d7f83fc12535f1a4d7b3e4a : lanelet extraction process to lambda function